### DISCUSSION TITLE
Routing with common prefix

### DISCUSSION THREAD

**Discussion by MichalSznajder** [dotnet/aspnetcore#58837] [Q&A] [answered]
Assume app with following routes:

```
app.MapGet("/first", () => "First");
app.MapGet("/{param}/second", () => "Second");
```

In this scenario accessing `/first/second` will result in response `Second`.

After some debugging I found following code in `DfaMarcher` 
https://github.com/dotnet/aspnetcore/blob/97c6f0df0d7b643dd5bac6f8a8641bd16738d2c5/src/Http/Routing/src/Matching/DfaMatcherBuilder.cs#L330-L336

This results with `DfaMatcher` table looking like this:

```
/first/ HTTP: GET -> 'HTTP: GET /first' 
/first/ HTTP: * -> '405 HTTP Method Not Supported' 
/first/second/ HTTP: GET -> 'HTTP: GET /{param}/second' 
/first/second/ HTTP: * -> '405 HTTP Method Not Supported' 
/{...}/second/ HTTP: GET -> 'HTTP: GET /{param}/second' 
/{...}/second/ HTTP: * -> '405 HTTP Method Not Supported' 
```

https://github.com/dotnet/aspnetcore/issues/38117 mentions 
> We've seen some issues with large DFAs as a result of some patterns that involve a large amount of variable prefixes. We've done work in 6.0 to address this via automatically trimming the DFA we generate by leveraging existing route constraints as well as the existing structure in complex segments.

so I tried 
```
app.MapGet("/first", () => "First");
app.MapGet("/{param:int}/second", () => "Second");
```
that results in `DfaTable` of:
```
/first/ HTTP: GET -> 'HTTP: GET /first' 
/first/ HTTP: * -> '405 HTTP Method Not Supported' 
/{...}/second/ HTTP: GET -> 'HTTP: GET /{param}/second' 
/{...}/second/ HTTP: * -> '405 HTTP Method Not Supported' 
```

However I would like `param` to be anything and `/first/second` return "First".

Is there a way to change it in automatic way?

**Comment by gfoidl** [comment, marked as answer]
@MichalSznajder I created an issue https://github.com/dotnet/aspnetcore/issues/58838 for this. Let's continue over there.

Thanks for raising that concern anyway!
